torch/_inductor/codegen/cpp_bmm_template.py

```diff
@@ -0,0 +1,249 @@
+# mypy: allow-untyped-defs
+import contextlib
+from typing import Any, Callable, Dict, List, Optional, Tuple
+from unittest.mock import patch
+
+import sympy
+
+from .. import ir, lowering as L
+from ..select_algorithm import PartialRender
+from ..virtualized import V
+from .cpp_gemm_template import (
+    CppGemmTemplate,
+    GEMM_TEMPLATE,
+    get_padded_n,
+    MICROKERNEL_DEF,
+)
+from .cpp_micro_gemm import LayoutType
+from .cpp_template_kernel import CppTemplateKernel
+from .cpp_utils import DTYPE_TO_CPP, GemmBlocking
+
+
+GEMM_SINGLE_THREAD_MM_STUB = r"""
+{{kernel.def_kernel(
+    inputs={"X": X, "W": W},
+    outputs={"Y": Y},
+    aliases=aliases,
+    function_name="single_thread_mm",
+    symbols=[b_index],
+    placeholder="<SINGLE_THREAD_DEF_MM_FOR_BMM>")}}"""
+
+GEMM_THREADED_MM_STUB = r"""
+{{kernel.def_kernel(
+    inputs={"X": X, "W": W},
+    outputs={"Y": Y},
+    aliases=aliases,
+    function_name="threaded_mm",
+    symbols=[b_index],
+    placeholder="<THREADED_MM_DEF_FOR_BMM>")}}"""
+
+BMM_WRAPPER = r"""
+extern "C"
+{{kernel.def_kernel(inputs={"X": BX, "W": BW}, outputs={"Y": BY}, aliases=aliases)}}
+{
+    const int64_t B = {{kernel.size(BY_2d, 0)}};
+    {%- if num_threads > 1 %}
+    constexpr int64_t num_threads = {{num_threads}};
+    int64_t B_single_thread_block = (B / num_threads) * num_threads;
+
+    #pragma omp parallel for num_threads({{num_threads}})
+    {%- else %}
+    int64_t B_single_thread_block = B;
+    {%- endif %}
+    for (int64_t b_start = 0; b_start < B_single_thread_block; ++b_start) {
+        {{template.get_gemm_function_call(
+            kernel,
+            "single_thread_mm",
+            "<SINGLE_THREAD_CALL_FOR_BMM>",
+            b_index="b_start",
+        )}}
+    }
+    for (int64_t b_start = B_single_thread_block; b_start < B; ++b_start) {
+        {{template.get_gemm_function_call(
+            kernel,
+            "threaded_mm",
+            "<THREADED_MM_CALL_FOR_BMM>",
+            b_index="b_start",
+        )}}
+    }
+}
+"""
+
+
+class CppBmmTemplate(CppGemmTemplate):
+    def __init__(
+        self,
+        input_nodes,
+        layout: ir.Layout,
+        num_threads: int,
+        register_blocking: GemmBlocking,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+        name="bmm",
+    ):
+        super().__init__(
+            input_nodes,
+            layout,
+            num_threads,
+            register_blocking,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            epilogue_creator=epilogue_creator,
+            name=name,
+        )
+        # Value may be changed after micro_gemm is instantiated if using VNNI layout
+        self.should_block_weights = False
```

Comment:
Feels we can directly check `micro_gemm.b_layout` instead of introducing another flag.

Reply:
Using blocked layout doesn't necessarily mean that weights are packed in VNNI layout. Current code does both of these together as part of preprocessor, but down the road when VNNI-packing is done within the template, then blocked layout and VNNI-backed layout need separate flags.